File: stone_lib/src/main/java/com/github/klee0kai/stone/_hidden_/types/holders/SingleItemHolder.java
package com.github.klee0kai.stone._hidden_.types.holders;

import com.github.klee0kai.stone._hidden_.types.ListUtils;
import com.github.klee0kai.stone._hidden_.types.SwitchCacheParams;
import com.github.klee0kai.stone.wrappers.Ref;

import java.lang.ref.Reference;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicInteger;

import static com.github.klee0kai.stone._hidden_.types.holders.StoneRefType.ListObject;
import static com.github.klee0kai.stone._hidden_.types.holders.StoneRefType.StrongObject;

/**
 * Stone Private class
 */
public class SingleItemHolder<T> {

    private final StoneRefType defType;
    private StoneRefType curRefType;

    private Object refHolder = null;
    private final AtomicInteger shedTaskCount = new AtomicInteger(0);

    public SingleItemHolder(StoneRefType defType) {
        this.defType = defType;
        this.curRefType = defType;
    }

    public synchronized T get() {
        switch (curRefType) {
            case StrongObject:
                return (T) refHolder;
            case WeakObject:
            case SoftObject:
                Reference<T> ref = ((Reference<T>) refHolder);
                return ref != null ? ref.get() : null;
            default:
                return null;
        }
    }

    public synchronized List<T> getList() {
        switch (curRefType) {
            case ListObject:
                return (List<T>) refHolder;
            case ListWeakObject:
            case ListSoftObject:
                return ListUtils.format((List<Reference<T>>) refHolder, Reference::get);
            default:
                return null;
        }
    }


    public synchronized void set(Ref<T> creator, boolean onlyIfNull) {
        if (Objects.equals(curRefType, StrongObject)) {
            if (refHolder != null && onlyIfNull) return;
            refHolder = creator.get();
            return;
        }
        ListUtils.IFormat<T, Reference<T>> formatter = curRefType.formatter();
        if (formatter == null) return;
        if (!onlyIfNull) {
            //switch ref type case
            refHolder = formatter.format(creator.get());
            return;
        }

        Reference<T> ref = (Reference<T>) refHolder;
        if (ref == null || ref.get() == null) {
            refHolder = formatter.format(creator.get());
        }
    }

    public synchronized void setList(Ref<List<T>> creator, boolean onlyIfNull) {
        if (Objects.equals(curRefType, ListObject)) {
            if (!onlyIfNull || refHolder == null) {
                refHolder = creator.get();
                return;
            }

            // init nulls if needed
            List<T> created = null;
            List<T> list = (List<T>) refHolder;
            for (int i = 0; i < list.size(); i++) {
                if (list.get(i) == null) {
                    if (created == null) created = creator.get();
                    list.set(i, created.get(i));
                }
            }
            return;
        }
        ListUtils.IFormat<T, Reference<T>> formatter = curRefType.formatter();
        if (formatter == null) return;
        List<Reference<T>> refList = (List<Reference<T>>) refHolder;
        if (refList == null || !onlyIfNull) {
            refHolder = ListUtils.format(creator.get(), formatter);
            return;
        }

        // init nulls if needed
        List<T> created = null;
        for (int i = 0; i < refList.size(); i++) {
            if (refList.get(i) == null || refList.get(i).get() == null) {
                if (created == null) created = creator.get();
                refList.set(i, formatter.format(created.get(i)));
            }
        }
    }

    public synchronized void setRefType(StoneRefType refType) {
        if (curRefType == refType) return;
        if (defType.isList()) {
            List<T> ob = getList();
            curRefType = refType.forList();
            setList(() -> ob, false);
        } else {
            T ob = get();
            curRefType = refType.forSingle();
            set(() -> ob, false);
        }
    }


    public synchronized void reset() {
        curRefType = defType;
        refHolder = null;
    }


    public synchronized void switchCache(SwitchCacheParams args) {
        switch (args.cache) {
            case Default:
                setRefType(defType);
                return;
            case Reset:
                reset();
                return;
            case Weak:
                setRefType(StoneRefType.WeakObject);
                break;
            case Soft:
                setRefType(StoneRefType.SoftObject);
                break;
            case Strong:
                setRefType(StoneRefType.StrongObject);
                break;
        }

        if (args.time > 0) {
            shedTaskCount.incrementAndGet();
            args.scheduler.schedule(new ScheduleTask(args.time) {
                @Override
                public void run() {
                    if (shedTaskCount.decrementAndGet() <= 0) setRefType(defType);
                }
            });
        }
    }

}

File: stone_multiplatform/src/commonMain/kotlin/com/github/klee0kai/stone/__hidden__/types/holders/SingleItemHolder.kt
package com.github.klee0kai.stone.__hidden__.types.holders

import com.github.klee0kai.stone.__hidden__.StoneScope
import com.github.klee0kai.stone.__hidden__.SwitchCacheParam
import com.github.klee0kai.stone.annotations.component.SwitchCache
import com.github.klee0kai.stone.weakref.Ref
import kotlinx.atomicfu.atomic
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlin.time.Duration

/**
 * Stone Private class
 */
@Suppress("UNCHECKED_CAST")
class SingleItemHolder<T>(
    private val defType: StoneRefType
) {
    private var curRefType: StoneRefType = defType

    private var refHolder: Any? = null
    private val shedTaskCount = atomic(0)


    fun get(): T? = when (curRefType) {
        StoneRefType.StrongObject -> refHolder as T?
        StoneRefType.WeakObject, StoneRefType.SoftObject -> (refHolder as Ref<T?>?)?.get()
        else -> null
    }

    fun getList(): List<T?>? = when (curRefType) {
        StoneRefType.ListObject -> refHolder as MutableList<T>?
        StoneRefType.ListWeakObject, StoneRefType.ListSoftObject -> (refHolder as MutableList<Ref<T?>?>?)?.map { it?.get() }
        else -> null
    }


    fun set(
        creator: Ref<T?>,
        onlyIfNull: Boolean,
    ) {
        if (curRefType == StoneRefType.StrongObject) {
            if (refHolder != null && onlyIfNull) return
            refHolder = creator.get()
            return
        }
        val formatter: (T?) -> Ref<T?> = curRefType.formatter<T>() ?: return
        if (!onlyIfNull) {
            //switch ref type case
            refHolder = formatter(creator.get())
            return
        }

        val ref: Ref<T?>? = refHolder as Ref<T?>?
        if (ref == null || ref.get() == null) {
            refHolder = formatter(creator.get())
        }
    }

    fun setList(
        creator: Ref<List<T?>?>,
        onlyIfNull: Boolean,
    ) {
        if (curRefType == StoneRefType.ListObject) {
            if (!onlyIfNull || refHolder == null) {
                refHolder = creator.get()?.toMutableList()
                return
            }

            // init nulls if needed
            var created: List<T?>? = null
            val list = refHolder as MutableList<T?>
            for (i in list.indices) {
                if (list[i] == null) {
                    if (created == null) created = creator.get()
                    list[i] = created?.get(i)
                }
            }
            return
        }

        val formatter = curRefType.formatter<T>() ?: return
        val refList = refHolder as? MutableList<Ref<T?>?>?
        if (!onlyIfNull || refList == null) {
            refHolder = creator.get()?.map { formatter(it) }?.toMutableList()
            return
        }

        // init nulls if needed
        var created: List<T?>? = null
        for (i in refList.indices) {
            if (refList[i] == null || refList[i]?.get() == null) {
                if (created == null) created = creator.get()
                refList[i] = formatter(created?.get(i))
            }
        }
    }

    fun setRefType(
        refType: StoneRefType,
    ) {
        if (curRefType == refType) return
        if (defType.isList) {
            val ob = this.getList()
            curRefType = refType.forList()
            setList(creator = Ref { ob }, false)
        } else {
            val ob = get()
            curRefType = refType.forSingle()
            set(creator = Ref { ob }, false)
        }
    }


    fun reset() {
        curRefType = defType
        refHolder = null
    }


    fun switchCache(args: SwitchCacheParam) {
        when (args.cache) {
            SwitchCache.CacheType.Default -> {
                setRefType(defType)
                return
            }

            SwitchCache.CacheType.Reset -> {
                reset()
                return
            }

            SwitchCache.CacheType.Weak -> setRefType(StoneRefType.WeakObject)
            SwitchCache.CacheType.Soft -> setRefType(StoneRefType.SoftObject)
            SwitchCache.CacheType.Strong -> setRefType(StoneRefType.StrongObject)
        }

        if (args.time != Duration.INFINITE) {
            shedTaskCount.incrementAndGet()
            StoneScope.stoneCoroutineScope.launch {
                delay(args.time)
                if (shedTaskCount.decrementAndGet() <= 0) {
                    setRefType(defType)
                }
            }
        }
    }
}
